core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java

```diff
@@ -75,6 +79,113 @@ public WriteResult complete() throws IOException {
         .build();
   }
 
+  /**
+   * Base delta writer to write both insert records and equality-deletes.
+   */
+  protected abstract class BaseDeltaWriter implements Closeable {
+    private RollingFileWriter dataWriter;
+    private RollingEqDeleteWriter eqDeleteWriter;
+    private SortedPosDeleteWriter<T> posDeleteWriter;
+    private StructLikeMap<PathOffset> insertedRowMap;
+
+    public BaseDeltaWriter(PartitionKey partition, Schema eqDeleteSchema) {
+      Preconditions.checkNotNull(eqDeleteSchema, "equality-delete schema could not be null.");
+
+      this.dataWriter = new RollingFileWriter(partition);
+
+      this.eqDeleteWriter = new RollingEqDeleteWriter(partition);
+      this.insertedRowMap = StructLikeMap.create(eqDeleteSchema.asStruct());
+
+      this.posDeleteWriter = new SortedPosDeleteWriter<>(appenderFactory, fileFactory, format, partition);
+    }
+
+    /**
+     * Make the generic data could be read as a {@link StructLike}.
+     */
+    protected abstract StructLike asStructLike(T data);
+
+    protected abstract StructLike asCopiedKey(T row);
```

Comment:
I think this could be removed from the API because this already has `asStructLike`. Rather than relying on the implementation to create the projection and copy it, this could be implemented here like this:

```java
  protected StructLike asCopiedKey(T row) {
    return structProjection.copy().wrap(asStructLike(row));
  }
```

We would need to create `StructProjection` in this class, but it would be fairly easy.

Reply:
I think the above implementation of `asCopiedKey` may has problems, because in some compute engine --- for example flink,  it will use the singleton `RowDataWrapper` instance to implement the `asStructLike`,  then for both the old key and copied key, they are sharing the same `RowDataWrapper` which has wrapped the new `RowData`.   That is messing up the keys in `insertedRowMap`.

Considered this issue,  I finally decided to abstract a separate `asCopiedKey` to clone a totally different key which won't share object values with the old one.

Reply:
Good catch. I like the idea of copying the data into a new struct.